### PR TITLE
break module dependency on namespace

### PR DESCRIPTION
**Pull request by melvinw**

### CODE REVIEW

**Comment by melvinw** [comment]
currently writing tests for this. just submitting to get feedback on the general changes before going much further

File: core/module.h [hunk 1]
```diff
@@ -97,229 +101,221 @@ struct Command {
...
-
-  virtual Module *CreateModule(const std::string &name) const {
-    T *m = new T;
-    m->name_ = name;
-    m->mclass_ = this;
-    return m;
-  };
+    struct snobj *RunCommand(Module *m, const std::string &user_cmd,
+        struct snobj *arg) const {
+      for (auto &cmd : cmds_) {
```
Comment (barath) [inline core/module.h]: Unless you need to modify it, you might make this const auto &

File: core/module.cc [hunk 2]
```diff
@@ -101,104 +123,90 @@ static std::string set_default_name(const std::string &class_name,
...
-    if (!find_module(name.c_str())) return name;  // found an unallocated name!
+    if (!all_modules_.count(name)) return name;
   }
 
   promise_unreachable();
 }
 
-static int register_module(Module *m) {
-  int ret;
+std::map<std::string, ModuleBuilder> &ModuleBuilder::all_module_builders_holder(bool reset) {
```
Comment (barath) [inline core/module.cc]: Might add a comment about this method.

File: core/module.cc [hunk 3]
```diff
@@ -47,38 +34,73 @@ task_id_t task_to_tid(struct task *t) {
...
+void ModuleBuilder::DestroyAllModules() {
+  int ret;
+  for (auto it = all_modules_.begin(); it != all_modules_.end(); ++it) {
+    ret = DestroyModule(it->second, false);
+    if (ret) {
+      LOG(ERROR) << "Error destroying module '" << it->first << "' (errno = " << ret << ")";
+      ++it;
+      continue;
+    }
+    all_modules_.erase(it);
```
Comment (melvinw) [inline core/module.cc]: woah. actually kinda surprised this works. will restructure

Comment (barath) [inline core/module.cc]: Since you do the erase in this method, you could do a postincrement on the iterator.

**Comment by melvinw** [comment]
thanks for the comments. resubmitting after cleaning things up and writing tests.

**Pull request closed without merging**